### DISCUSSION TITLE
Execution Order for JUnit Tests

### DISCUSSION THREAD

**Discussion by GreihMurray** [junit-team/junit-framework#2948] [Q&A] [answered]
Hey Im just really curious, why order the JUnit tests execution by hashing the names and ordering that way? Why not just start at the top of the file and work your way down? Could someone just explain this decision?

**Comment by sbrannen** [comment, marked as answer]
JUnit introspects the methods of a compiled test class using Java's reflection APIs. JUnit is not aware of the source code of test classes.

The JDK does not provide any API that allows one to infer the order in which methods are declared in the source code.

The only built-in tool we have is `java.lang.Class.getDeclaredMethods()` which makes no guarantee about the ordering of the methods.

Thus, in order to enforce consistent ordering of test methods across multiple runs of the same test suite, we implemented a mechanism that is non-obvious yet deterministic, as stated in the [User Guide](https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-execution-order):

> By default, test classes and methods will be ordered using an algorithm that is deterministic but intentionally non-obvious. This ensures that subsequent runs of a test suite execute test classes and test methods in the same order, thereby allowing for repeatable builds.

**Reply by GreihMurray** [reply]
Thank you very much for the reply, that makes so much more sense now
